### PR TITLE
Introduce ChunkShardingSpec as a model sharding specification.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58517 ShardedTensor framework for ChunkedShardingSpec
* #57409 Add GenericShardingSpec for generic tensor sharding.
* **#55728 Introduce ChunkShardingSpec as a model sharding specification.**

Full design: https://github.com/pytorch/pytorch/issues/55207

This PR introduces ChunkShardingSpec (SingleShardingSpec in the design). Used
the name ChunkShardingSpec since it is very similar to `torch.chunk` in terms
of how a Tensor is split up and feels more clear compared to SingleShardingSpec.

Differential Revision: [D27694108](https://our.internmc.facebook.com/intern/diff/D27694108/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27694108/)!

### CODE REVIEW

File: torch/distributed/utils.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import torch
+
+def _parse_remote_device(remote_device: str):
```
Comment (SciPioneer) [inline torch/distributed/utils.py:3]: I understand that compared with `_parse_remote_device` in `torch/distributed/rpc/utils.py`, here we also want to support another format `rank:<rank>/<device>`. Shall we unify these two very similar functions?

File: torch/distributed/utils.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    Parses the remote device.
+
+    Args:
+        remote_device (str): Device on the destination worker where we'd like to place this module.
+            The format should be one of the following:
+
+                1. "<workername>/<device>", where the device field can be parsed as torch.device type.
+                   E.g., "trainer0/cpu", "trainer0", "ps0/cuda:0".
+                   In addition, the device field can be optional and the default value is "cpu".
+                2. "rank:<rank>/<device>", where <rank> is the rank of the
```
Comment (SciPioneer) [inline torch/distributed/utils.py:14]: Thanks for unifying the util functions for both sharded tensor and remote module! This means that wow we are also supporting the second format for `RemoteModule`. Is there any documentation that we should update for `RemoteModule`?

File: torch/distributed/_sharding_spec/api.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            dimension or a string in case of named tensors where dimensions are
+            named.
+        placement(List[Device] or List[PlacementSpec]):
+            Specifies the placement of each shard of the Tensor. The size of
+            the list represents the number of shards to be created. This
+            parameter can be a list of devices
+            (ex: ["rank:0/cuda:0", "rank:1/cuda:1"]) or a list of custom
+            placement specs.
+
+            The device can be a local device or a remote device specified by one
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:58]: Not against this functionality for supporting sharding a tensor across remote devices. Out of curiosity, have you seen any performance gain by such across-node tensor sharding? It seems that this will increase a lot of communication overheads.

File: torch/distributed/_sharding_spec/_internals.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import torch
+from torch.distributed.utils import _parse_remote_device
+
+def is_valid_device(device):
+    """
+    Checks if this is a valid remote/local device.
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/_internals.py:6]: Ni: Change the order of "remote/local" here to be consistent with the implementation?

File: torch/distributed/_sharding_spec/api.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        self._placement = placement
+
+    @staticmethod
+    def _verify_devices(placement):
+        for dev in placement:
+            if not isinstance(dev, PlacementSpec) and not is_valid_device(dev):
+                raise ValueError(f'{dev} is not a valid device')
+
+    @staticmethod
+    def _verify_dim(dim):
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:82]: Shall we also very that the value of `dim` is in the range of the tensor's dimensionality? It cannot be negative or a very large value.

File: test/distributed/_sharding_spec/test_sharding_spec.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import torch
+from torch.testing._internal.common_utils import TestCase
+from torch.distributed._sharding_spec import (
+    ChunkShardingSpec,
+    DevicePlacement
+)
+
+class TestShardingSpec(TestCase):
+
+    def test_device_placement(self):
```
Comment (SciPioneer) [inline test/distributed/_sharding_spec/test_sharding_spec.py:10]: Since you have already reused `_parse_remote_device` util function, it will be great to also share the duplicate test cases for validation (probably by sharing two const lists to represent valid and invalid inputs). We may even consider reusing `DevicePlacement` in `RemoteModule` class, as `DevicePlacement` is more generic than tensor sharding.

File: torch/distributed/utils.py (hunk 2 above)
Comment (SciPioneer) [reply torch/distributed/utils.py:14]: It may be even better to just reuse `DevicePlacement` class in `RemoteModule`, then we don't need to maintain two copies of valid formats.

File: torch/distributed/utils.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            on = int(fields[1])
+        else:
+            raise ValueError(
+                f"Could not parse remote_device: {remote_device}. The valid format is "
+                "'<workername>/<device>' or 'rank:<rank>/<device>"
+            )
+    elif len(fields) > 2:
+        raise ValueError(
+            f"Could not parse remote_device: {remote_device}. The valid format is "
+            "'<workername>/<device>' or 'rank:<rank>/<device>"
```
Comment (rohan-varma) [inline torch/distributed/utils.py:62]: nit: dedupe these error messages into a constant str?

File: test/distributed/_sharding_spec/test_sharding_spec.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        DevicePlacement("rank:0/cpu")
+        DevicePlacement("rank:0")
+
+        # invalid devices
+        with self.assertRaisesRegex(ValueError, "not a valid device"):
+            DevicePlacement("cuda:foo")
+        with self.assertRaisesRegex(ValueError, "not a valid device"):
+            DevicePlacement("foo:0")
+        with self.assertRaisesRegex(ValueError, "not a valid device"):
+            DevicePlacement("rank:0/cuda:foo")
```
Comment (SciPioneer) [inline test/distributed/_sharding_spec/test_sharding_spec.py:25]: Can we try a few more test cases?
1) "cuda:100"
2) "cpu2"

**Comment by SciPioneer** [comment]
Please fix the test failures.

File: torch/distributed/_sharding_spec/api.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        self._device = device
+
+    @property
+    def device(self) -> Device:
+        """
+        Retrieves the device for placement.
+        """
+        return self._device
+
+class ChunkShardingSpec(PlacementSpec):
```
Comment (divchenko) [inline torch/distributed/_sharding_spec/api.py:40]: ChunkShardingSpec , DevicePlacement and PlacementSpec - naming seems a bit off. Some have 'Placement', some have 'Spec' in their names. It's not obvious they belong to the same class hierarchy.

File: torch/distributed/utils.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        A workername/rank and a device.
+    """
+
+    PARSE_ERROR = (
+        f"Could not parse remote_device: {remote_device}. The valid format is "
+        "'<workername>/<device>' or 'rank:<rank>/<device>'"
+    )
+
+    fields = remote_device.split("/")
+    if len(fields) == 2:
```
Comment (mrshenli) [inline torch/distributed/utils.py:28]: Looks like this method only checks whether the format is correct. Do we also need to check whether the ranks are valid before it is used by, e.g., DevicePlacementSpec?

File: torch/distributed/utils.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        [on, device] = fields
+    elif len(fields) == 1:
+        on = fields[0]
```
Comment (mrshenli) [inline torch/distributed/utils.py:31]: ditto for checking the existence of worker names

File: torch/distributed/utils.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        device = "cpu"
+    else:
+        raise ValueError(PARSE_ERROR)
+
+    # Since the workername in the input remote device won't be validated until the created remote module is executed,
```
Comment (mrshenli) [inline torch/distributed/utils.py:36]: This means we expect ShardingSpec to be created before `init_process_group` and `init_rpc`? If so, does it mean users need to do another round of correctness verification after c10d/rpc are initialized? Will it make sense to require that ShardingSpec can only be created after c10d/rpc initialization, so that we can do more correctness checks here?

File: torch/distributed/utils.py (hunk 10 above)
Comment (pritamdamania87) [reply torch/distributed/utils.py:28]: The specification itself might be tied to different runtime environments. For example, you can store this specification with a checkpoint and then load it during checkpoint load time. As a result, I thought its better not to validate runtime environment here. For example, we don't verify GPU indexes either.

File: torch/distributed/utils.py (hunk 11 above)
Comment (pritamdamania87) [reply torch/distributed/utils.py:31]: same as above, I think this would be done at a higher level like when you actually create a ShardedTensor

File: torch/distributed/utils.py (hunk 12 above)
Comment (pritamdamania87) [reply torch/distributed/utils.py:36]: I don't think ShardingSpec creation is tied to `init_process_group` or `init_rpc` and it shouldn't be tied to them. The specification should be independent of the runtime environment such that I can apply the same specification to different training jobs as an example.

File: torch/distributed/utils.py (hunk 10 above)
Comment (mrshenli) [reply torch/distributed/utils.py:28]: I see. That means, we consider the sharding spec as counterparts of Device/RemoteDevice, and the check will be enforced at the loading time?

Comment (pritamdamania87) [reply torch/distributed/utils.py:28]: Right that is what I had in mind.

File: torch/distributed/_sharding_spec/api.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    This is type of PlacementSpec that defines the placement as being sharded
+    across multiple devices. In particular, it represents sharding a Tensor
+    along a single dimension into equal chunks (similar to :meth:`torch.chunk`).
+
+    Args:
+        dim (int or str):
+            The dimension to shard on, could be an integer representing the
+            dimension or a string in case of named tensors where dimensions are
+            named.
+        placement(List[Device] or List[PlacementSpec]):
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:51]: Should `List[PlacementSpec]` be `List[DevicePlacementSpec]`? Otherwise, it means that a `ChunkShardingSpec` can be represented by `List[ChunkShardingSpec]`, which means that the sharding here can be hierarchical.

File: torch/distributed/_sharding_spec/api.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+            placement specs.
+
+            The device can be a local device or a remote device specified by one
+            of the following remote formats:
+
+                1. "rank:<rank>/<device>" (ex: "rank:0/cuda:0").
+                2. "<worker_name>/<device>" (ex: "trainer0/cuda:0").
+    """
+
+    ShardPlacement = Union[List[Device], List[PlacementSpec]]
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:65]: Optional: `List[Union[Device, PlacementSpec]]` can be even more flexible if you want.

Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:65]: Better to use the plural form. Otherwise I sounds more like the placement of a single shard.

File: torch/distributed/_sharding_spec/api.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    ShardingDim = Union[int, str]
+
+    def __init__(self, dim: ShardingDim, placement: ShardPlacement):
+        super(ChunkShardingSpec, self).__init__()
+        self._verify_dim(dim)
+        self._verify_devices(placement)
+        self._dim = dim
+        self._placement = placement
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:73]: Ditto. `_placement` -> `_shard_placements`

File: torch/distributed/_sharding_spec/api.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
+from abc import ABC
+import torch
+from typing import List, Union
+
+from ._internals import is_valid_device
+
+Device = Union[torch.device, int, str]
+
+class PlacementSpec(ABC):
```
Comment (SciPioneer) [inline torch/distributed/_sharding_spec/api.py:9]: Why do we need a common superclass here? Will there be any major common attributes or methods?

I feel `List[DevicePlacementSpec]` should just be a data member of `ChunkShardingSpec`, and the relationship is only composition and little inheritance.

File: torch/distributed/utils.py (hunk 10 above)
Comment (SciPioneer) [reply torch/distributed/utils.py:28]: Agree. This syntax-only non-semantics check is consistent with the remote device validation in RPC framework.

File: torch/distributed/_sharding_spec/api.py (hunk 16 above)
Comment (pritamdamania87) [reply torch/distributed/_sharding_spec/api.py:9]: `DevicePlacementSpec` would be useful in the case where we want to shard an nn.Module. Let's say we have 3 tensors in the module and we want to place them on three different devices. `DevicePlacementSpec` can be used in conjunction with "Composite ShardingSpec" from https://github.com/pytorch/pytorch/issues/55207 to address this.

File: torch/distributed/_sharding_spec/api.py (hunk 13 above)
Comment (pritamdamania87) [reply torch/distributed/_sharding_spec/api.py:51]: I kept it `List[PlacementSpec]` for extra flexibility. For example you could subclass PlacementSpec to specify your own custom placement (ex: some database like rocksdb).

Comment (SciPioneer) [reply torch/distributed/_sharding_spec/api.py:51]: Can you add a comment about this? So powerful users know that they can subclass `PlacementSpec` for this purpose.

File: torch/distributed/_sharding_spec/api.py (hunk 16 above)
Comment (SciPioneer) [reply torch/distributed/_sharding_spec/api.py:9]: I see, so "sharding" here means not only intra-layer sharding, but also inter-layer sharding (where `DevicePlacementSpec`, i.e., no intra-layer sharding, can be applied).

**Review by wayi1 — APPROVED** [review]
